### ISSUE TITLE
[[INVALID],INVALID] FORCE-TERMINATE AT Not found:-13 - error plm_rsh_component.c(327)

### ISSUE THREAD

**Issue by adrianreber** [open-mpi/ompi#7330]
While trying to build petsc for OpenHPC I get a core dump in mpiexec on CentOS 8 with the Open MPI 4.0.2 build from OpenHPC:

```
[   91s] + mpiexec --oversubscribe -n 1 printenv
[   91s] --------------------------------------------------------------------------
[   91s] The value of the MCA parameter "plm_rsh_agent" was set to a path
[   91s] that could not be found:
[   91s] 
[   91s]   plm_rsh_agent: rsh
[   91s] 
[   91s] Please either unset the parameter, or check that the path is correct
[   91s] --------------------------------------------------------------------------
[   91s] [ohpc1:04221] [[INVALID],INVALID] FORCE-TERMINATE AT Not found:-13 - error plm_rsh_component.c(327)
[   91s] [ohpc1:04221] *** Process received signal ***
[   91s] [ohpc1:04221] Signal: Segmentation fault (11)
[   91s] [ohpc1:04221] Signal code: Address not mapped (1)
[   91s] [ohpc1:04221] Failing at address: (nil)
[   91s] [ohpc1:04221] [ 0] /lib64/libpthread.so.0(+0x12dd0)[0x7f70eb00add0]
[   91s] [ohpc1:04221] *** End of error message ***
```
At first I was not able to reproduce it outside of our build system, but setting `OMPI_MCA_plm_rsh_agent` to `does-not-exist` I am able to reproduce this with the OS provided Open MPI 4.0.2 on RHEL 8:
```
[mpi@sms ~]$ export OMPI_MCA_plm_rsh_agent=does-not-exist
[mpi@sms ~]$ mpiexec --oversubscribe -n 1 printenv 
--------------------------------------------------------------------------
The value of the MCA parameter "plm_rsh_agent" was set to a path
that could not be found:

  plm_rsh_agent: does-not-exist

Please either unset the parameter, or check that the path is correct
--------------------------------------------------------------------------
[sms:16573] [[INVALID],INVALID] FORCE-TERMINATE AT Not found:-13 - error plm_rsh_component.c(327)
[sms:16573] *** Process received signal ***
[sms:16573] Signal: Segmentation fault (11)
[sms:16573] Signal code: Address not mapped (1)
[sms:16573] Failing at address: (nil)
[sms:16573] [ 0] /lib64/libpthread.so.0(+0x12dd0)[0x7f6fb4771dd0]
[sms:16573] *** End of error message ***
Segmentation fault (core dumped)
```

I can also see this on Fedora 31 with Open MPI 4.0.2.

**Comment by jjhursey** [comment]
It should be looking for both `ssh` and `rsh` by default [here](https://github.com/open-mpi/ompi/blob/v4.0.2/orte/mca/plm/rsh/plm_rsh_component.c#L181).

I've seen this error before when trying to launch mpirun in a container image that does not have either one of them installed. Does your build environment have them installed?

I've been meaning to look into a mode that recognizes that there are not going to be any remote agents to launch via PLM - so there is no need for ssh/rsh to be installed. Right now I don't know of a workaround if they are not installed (other than installing them).

**Comment by adrianreber** [comment]
The build environment does not have them installed. I 'fixed' it for me by installing it. I was just surprised by the core dump and wanted to report it.

**Comment by jjhursey** [comment]
Thanks for the report. I'll try to take a look in the near term to see if I can first have it fail without the core dump, and second not fail in this scenario.

**Comment by hppritcha** [comment]
@jjhursey do you think you may be able to put in  a simple fix to avoid a core dump in time for 4.0.3 release?

**Comment by jjhursey** [comment]
I'll investigate today to see if I can pin it down.

**Comment by jjhursey** [comment]
I have a fix posted to `master` in PR #7330 and verified that the same patch does work in the `v4.0.x` branch.

```
shell $ unset OMPI_MCA_plm_rsh_agent
shell $ mpirun -np 2 hostname
c712f6n01
c712f6n01
shell $ export OMPI_MCA_plm_rsh_agent=does-not-exist
shell $ mpirun -np 2 hostname
--------------------------------------------------------------------------
The value of the MCA parameter "plm_rsh_agent" was set to a path
that could not be found:

  plm_rsh_agent: does-not-exist

Please either unset the parameter, or check that the path is correct
--------------------------------------------------------------------------
[c712f6n01:127205] [[INVALID],INVALID] FORCE-TERMINATE AT Not found:-13 - error plm_rsh_component.c(335)
shell $ echo $?
243
shell $ export OMPI_MCA_plm_rsh_agent=
shell $ mpirun -np 2 hostname
c712f6n01
c712f6n01
shell $ docker run -it --rm -v $MPI_ROOT:$MPI_ROOT --env MPI_ROOT=$MPI_ROOT centos:8 bash
[root@b43b0668151d /]# cat /etc/redhat-release 
CentOS Linux release 8.1.1911 (Core) 
[root@b43b0668151d /]# export PATH=$MPI_ROOT/bin:$PATH
[root@b43b0668151d /]# export LD_LIBRARY_PATH=$MPI_ROOT/lib:$LD_LIBRARY_PATH
[root@b43b0668151d /]# mpirun --allow-run-as-root --np 2 hostname
--------------------------------------------------------------------------
The value of the MCA parameter "plm_rsh_agent" was set to a path
that could not be found:

  plm_rsh_agent: ssh : rsh

Please either unset the parameter, or check that the path is correct
--------------------------------------------------------------------------
[b43b0668151d:00016] [[INVALID],INVALID] FORCE-TERMINATE AT plm_rsh_component.c:336 - error Not found(-13)
[root@b43b0668151d /]# echo $?
243
[root@b43b0668151d /]# export OMPI_MCA_plm_rsh_agent=
[root@b43b0668151d /]# mpirun --allow-run-as-root --np 2 hostname
b43b0668151d
b43b0668151d
```

In the process of testing, I found a workaround for environments without `ssh` or `rsh`. Setting the `export OMPI_MCA_plm_rsh_agent=` to the empty string will cause the `rsh` PLM to disqualify and fallback on the `isolated` PLM which will allow you to launch locally only.

**Comment by jjhursey** [comment]
The fix for this has been merged into `master` and the `v4.0.x` streams (and ported over to PRRTE `master`). When you have a chance please give one of the nightly builds a try to confirm the fix on your end. Thanks for the bug report!
